cmd/tsbs_generate_queries/databases/siridb/devops.go

```diff
@@ -0,0 +1,175 @@
+package siridb
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/uses/devops"
+	"github.com/timescale/tsbs/query"
+)
+
+// Devops produces SiriDB-specific queries for all the devops query types.
+type Devops struct {
+	*devops.Core
+}
+
+// NewDevops makes an Devops object ready to generate Queries.
+func NewDevops(start, end time.Time, scale int) *Devops {
+	return &Devops{devops.NewCore(start, end, scale)}
+}
+
+// GenerateEmptyQuery returns an empty query.SiriDB
+func (d *Devops) GenerateEmptyQuery() query.Query {
+	return query.NewSiriDB()
+}
+
+func (d *Devops) getHostWhereWithHostnames(hostnames []string) string {
+	hostnameClauses := []string{}
+	for _, s := range hostnames {
+		hostnameClauses = append(hostnameClauses, fmt.Sprintf("`%s`", s))
+	}
+	combinedHostnameClause := strings.Join(hostnameClauses, "|")
+	return "(" + combinedHostnameClause + ")"
+}
+
+func (d *Devops) getHostWhereString(nhosts int) string {
+	hostnames := d.GetRandomHosts(nhosts)
+	return d.getHostWhereWithHostnames(hostnames)
+}
+
+func (d *Devops) getMetricWhereString(metrics []string) string {
+	metricsClauses := []string{}
+	for _, s := range metrics {
+		metricsClauses = append(metricsClauses, fmt.Sprintf("`%s`", s))
+	}
+	combinedMetricsClause := strings.Join(metricsClauses, "|")
+	return "(" + combinedMetricsClause + ")"
+}
+
+const goTimeFmt = "2006-01-02 15:04:05Z"
+
+// GroupByTime selects the MAX for numMetrics metrics in the group `cpu`,
+// per minute for nhosts hosts,
+// e.g. in psuedo-SQL:
+//
+// In case of 1 metric:
+// select max(1m) from (`groupHost1` | ...) & `groupMetric1` between 'time1' and 'time2' merge as 'max METRIC for (HOST | ... ) using max(1)
+//
+// In case of multiple metrics
+// NOTE: it is not possible to merge multiple hosts per metric for a list of
+// metrics, only for one metric. In this case all series for the provided hosts
+// and metrics are returned:
+//
+// select max(1m) from (`groupHost1` | ...) & (`groupMetric1` | ...) between 'time1' and 'time2'
+func (d *Devops) GroupByTime(qi query.Query, nHosts, numMetrics int, timeRange time.Duration) {
+	interval := d.Interval.RandWindow(timeRange)
+	metrics := devops.GetCPUMetricsSlice(numMetrics)
+	whereMetrics := d.getMetricWhereString(metrics)
+	whereHosts := d.getHostWhereString(nHosts)
+
+	humanLabel := fmt.Sprintf("SiriDB %d cpu metric(s), random %4d hosts, random %s by 1m", numMetrics, nHosts, timeRange)
+	humanDesc := fmt.Sprintf("%s: %s", humanLabel, interval.StartString())
+
+	var siriql string
+	if numMetrics == 1 {
```

Comment:
I find this part a bit weird, but I haven't dived into the documentation for the sql query language. I'm a bit curious why the case of 1 metric would be more verbose than multiple metrics.

Reply:
The reason for this lies in how the time series are stored. We have stored the series per metric and per host. It is possible to return the series of e.g. 10 metrics  and 5 hosts (using regular expressions), this will return 50 time series in total. If we then merge these series, these 50 series are merged to 1 series. So merged over both host and metric. Unfortunately it is not possible yet to merge per metric for multiple metrics within one query request. Therefore if we want to merge only over host, we have to select one metric at a time. In cases like ` single-groupby-5-8-1 ` we chose to return all the series separate instead and no merging is done.